### PR TITLE
SelfProtocore upsert validator

### LINKED ISSUE #179: Implement Protocore::upsert()
1. Implement `Protocore::upsert()` external function
    - Checks the reputation of the validator
    - We check 
        - if the validator is already there and has a +ve reputation, then no change
        - if the validator is not there and has a +ve reputation, then it is joining in the next metablock height
        - if the validator's reputation is zero, then it is logging out
        - (and other cases)
    - Proposal
    ```solidity
    function upsert(
        address _validator,
        uint256 _height,
        uint256 _reputation
    )
        external
    ```

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR contains below changes :
1. Added `upsertValidator` method in SelfProtocore contract.
2. Unit test case.

Upsert behaves in the following way:
1. It inserts validator if not already present and reputation is greater than 0.
2. It removes validator if it is already present and reputation is 0.

fixes #179

### CODE REVIEW

File: contracts/protocore/Protocore.sol [hunk 1]
```diff
@@ -176,6 +176,39 @@ contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
...
+        onlyCoconsensus
+    {
+        if(ValidatorSet.inValidatorSet(_validator, _height)) {
+            if(_reputation == 0) {
+                removeValidatorInternal(_validator, _height);
+            }
+        }
+        else {
+            if(_reputation > 0) {
+                insertValidatorInternal(_validator, _height);
```
Comment (deepesh-kn) [inline contracts/protocore/Protocore.sol]: For the past kernel heights, a validator can be inserted? 
I think that checks related to open kernel height are missing.

I will refer again to the comments from this PR https://github.com/mosaicdao/mosaic-1/pull/212

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Please see the comments, I think that a check related to kernel height is needed.

As a side note:
I think that this code should be in `SelfProtocore` and not `Protocore`.
`SelfProtocore` and `OriginProtocore` inherits the `Protocore` contract.
As per the specifications, `OriginProtocore` will not have validator sets and will rely on the `SelfProtocore` for any validator related queries. So the `SelfProtocore` will inherit `ValidatorSet` and not `Protocore`. We can discuss this point, and if agreed we can move this code to `SelfProtocore`.

File: contracts/protocore/Protocore.sol (hunk 1 above)
Comment (gulshanvasnani) [reply contracts/protocore/Protocore.sol]: Added check so that validator can enter only at current open kernel height.

File: contracts/validator/ValidatorSet.sol [hunk 2]
```diff
@@ -126,6 +128,11 @@ contract ValidatorSet {
             validatorBeginHeight[_validator] == 0 && validatorEndHeight[_validator] == 0
         );
 
+        require(
+            _beginHeight == _openKernelHeight,
```
Comment (deepesh-kn) [inline contracts/validator/ValidatorSet.sol]:
```suggestion
            _beginHeight >= _openKernelHeight,
```

File: contracts/validator/ValidatorSet.sol [hunk 3]
```diff
@@ -126,6 +128,11 @@ contract ValidatorSet {
             validatorBeginHeight[_validator] == 0 && validatorEndHeight[_validator] == 0
```
Comment (deepesh-kn) [inline contracts/validator/ValidatorSet.sol]: I still think that this should be `requires` and not `asserts`

File: test/self_protocore/upsert_validator.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
+// Copyright 2019 OpenST Ltd.
```
Comment (deepesh-kn) [inline test/self_protocore/upsert_validator.js]:
```suggestion
// Copyright 2020 OpenST Ltd.
```

File: test/self_protocore/upsert_validator.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        validator.reputation,
+        {
+          from: config.coconsensusAddress,
+        },
+      );
+
+      assert.notStrictEqual(
+        await config.selfProtocore.validators.call(validator.address),
+        Utils.NULL_ADDRESS,
+        'Expected address is SENTINEL_ADDRESS(0x1)',
```
Comment (deepesh-kn) [inline test/self_protocore/upsert_validator.js:107]: When we know what should be the expected result, then why don't we do a strict check?
Either change it to `strictEqual` or change the reason that it should not be a null address.

File: contracts/protocore/SelfProtocore.sol [hunk 6]
```diff
@@ -69,5 +69,43 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
...
+        address _validator,
+        uint256 _height,
+        uint256 _reputation
+    )
+        external
+        onlyCoconsensus
+    {
+        if(ValidatorSet.inValidatorSet(_validator, _height)) {
+            if(_reputation == 0) {
+                removeValidatorInternal(_validator, _height);
```
Comment (deepesh-kn) [inline contracts/protocore/SelfProtocore.sol:103]: Please add `openKernelHeight` for remove validator too.
`validatorBeginHeight[_validator] <= openKernelHeight,`

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Almost done, just few comments please check.

File: contracts/validator/ValidatorSet.sol (hunk 3 above)
Comment (gulshanvasnani) [reply contracts/validator/ValidatorSet.sol]: It was done because of the discussion at https://github.com/mosaicdao/mosaic-1/pull/212#discussion_r370237960

File: test/self_protocore/upsert_validator.js (hunk 5 above)
Comment (gulshanvasnani) [reply test/self_protocore/upsert_validator.js:107]: Correct. We know the expected result.
But it is done as per the discussion at https://github.com/mosaicdao/mosaic-1/pull/212#discussion_r370044049

File: contracts/protocore/SelfProtocore.sol (hunk 6 above)
Comment (gulshanvasnani) [reply contracts/protocore/SelfProtocore.sol:103]: I have added this check.

File: contracts/validator/ValidatorSet.sol (hunk 2 above)
Comment (gulshanvasnani) [reply contracts/validator/ValidatorSet.sol]: Done.

File: test/self_protocore/upsert_validator.js (hunk 4 above)
Comment (gulshanvasnani) [reply test/self_protocore/upsert_validator.js]: Done

File: contracts/protocore/SelfProtocore.sol [hunk 7]
```diff
@@ -69,5 +69,43 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
...
+     */
+    function upsertValidator(
+        address _validator,
+        uint256 _height,
+        uint256 _reputation
+    )
+        external
+        onlyCoconsensus
+    {
+        if(ValidatorSet.inValidatorSet(_validator, _height)) {
```
Comment (deepesh-kn) [inline contracts/protocore/SelfProtocore.sol]: Somehow these conditions are bothering me.
It should check if the validator set already existed irrespective of height provided.
Similarly, it should check if the end height is MAX_FUTURE_END_HEIGHT in order to remove the validator.

Comment (gulshanvasnani) [reply contracts/protocore/SelfProtocore.sol]: Thank you.
Added check to verify if validator existed irrespective of it's height.
Regarding check to verify end heigh of validator is MAX_FUTURE_END_HEIGHT, it's already happening in `ValidatorSet::removeValidatorInternal`.
https://github.com/mosaicdao/mosaic-1/blob/develop/contracts/validator/ValidatorSet.sol#L157

Comment (deepesh-kn) [reply contracts/protocore/SelfProtocore.sol]: I was suggesting this because I wanted to make sure that if a validator is removed once, then calling it again will not revert. Our discussion was that Upsert should not revert. Maybe this will not happen. I will need a second eye to check this logic. cc: @pgev

File: contracts/protocore/SelfProtocore.sol [hunk 8]
```diff
@@ -69,5 +69,43 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
             genesisAuxiliaryTargetBlockHash,
             genesisAuxiliaryTargetBlockNumber
         );
+        ValidatorSet.setupValidatorSet();
     }
+
+
+    /* External */
```
Comment (pgev) [inline contracts/protocore/SelfProtocore.sol]: `External Functions`

Comment (gulshanvasnani) [reply contracts/protocore/SelfProtocore.sol]: Done

**Comment by benjaminbollen** [comment]
Im OK with moving `Protocore::upsert` to `SelfProtocore::upsert`, but let's take note (again, escalation from same point) that the specification was incorrect, and mid-sprint deviations should be tracked (@jayeshbairagi can you keep an overview for retrospective? )

I do not agree that an extra `height` should to be passed to `insertValidatorInternal` or `removeValidatorInternal` (and changing the interface). In the validator count ticket there will be a better approach for stronger restricting insert and remove.

File: contracts/protocore/SelfProtocore.sol [hunk 9]
```diff
@@ -69,5 +69,42 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
...
+     */
+    function upsertValidator(
+        address _validator,
+        uint256 _height,
+        uint256 _reputation
+    )
+        external
+        onlyCoconsensus
+    {
+        if(validators[_validator] != address(0)) {
```
Comment (benjaminbollen) [inline contracts/protocore/SelfProtocore.sol]: doesn't this respect the interface better (aka `validators[]` should probably just be private and an `external view` function iterates over (some length) to return an array):

```
if (isInValidatorSet(_validator, openKernelHeight)) {...}
```
IMPORTANT : I updated the parameter to `openKernelHeight` in the suggestion above

Comment (benjaminbollen) [reply contracts/protocore/SelfProtocore.sol]: we can add an `assert(_height == openKernelHeight.add(1))`

File: contracts/validator/ValidatorSet.sol [hunk 10]
```diff
@@ -114,10 +114,12 @@ contract ValidatorSet {
      *
      * @param _validator Validator address.
      * @param _beginHeight Begin height for the validator.
+     * @param _openKernelHeight Current open kernel height.
```
Comment (benjaminbollen) [inline contracts/validator/ValidatorSet.sol]: this height should not be passed in, nor the new requirement be put.  Please revert this for `insertValidatorInternal` and `removeValidatorInternal`

**Review by benjaminbollen — CHANGES_REQUESTED** [review]
revert the changes to `insertValidatorInternal` and `removeValidatorInternal`

File: contracts/protocore/SelfProtocore.sol (hunk 9 above)
Comment (gulshanvasnani) [reply contracts/protocore/SelfProtocore.sol]: Thank you.
Added check `isInValidatorSet` and suggested `assert` before calling `insertValidatorInternal`. 
Yes, I liked your suggestion to make validators private and have an iterator function. Shall I create the ticket to address it ?

File: contracts/validator/ValidatorSet.sol (hunk 10 above)
Comment (gulshanvasnani) [reply contracts/validator/ValidatorSet.sol]: Reverted the change for `insertValidatorInternal` and `removeValidatorInternal`.

**Comment by gulshanvasnani** [comment]
I have reverted the changes in `insertValidatorInternal` and `removeValidatorInternal` methods.
Now, This PR is available for review.

File: contracts/protocore/SelfProtocore.sol [hunk 11]
```diff
@@ -69,5 +69,42 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
...
+     * @param _reputation Validator's reputation value.
+     */
+    function upsertValidator(
+        address _validator,
+        uint256 _height,
+        uint256 _reputation
+    )
+        external
+        onlyCoconsensus
+    {
```
Comment (benjaminbollen) [inline contracts/protocore/SelfProtocore.sol:99]: add `assert(_height == openKernelHeight.add(1));`

Comment (benjaminbollen) [reply contracts/protocore/SelfProtocore.sol:99]: specifically, move up L104

File: test/self_protocore/upsert_validator.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    config.coconsensusAddress = accountProvider.get();
+
+    config.epochLength = new BN(100);
+
+    config.genesisKernelHeight = new BN(1);
+    config.genesisKernelHash = Utils.getRandomHash();
+
+    config.core = accountProvider.get();
+    config.genesis = {};
+    config.genesis.auxiliaryParentVoteMessageHash = Utils.getRandomHash();
```
Comment (benjaminbollen) [inline test/self_protocore/upsert_validator.js:40]: this should either move into a local `utils.js`, or we can have `SelfProtocoreValidatorSetMock` that bypasses these params

File: test/self_protocore/upsert_validator.js [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+      Utils.getRandomHash(),
+      new BN(100),
+      new BN(20),
+      {
+        from: config.coconsensusAddress,
+      },
+    );
+  });
+
+  contract('Positive Tests', async () => {
```
Comment (benjaminbollen) [inline test/self_protocore/upsert_validator.js:93]: there are definitely more unit tests valuable here for the different edge-cases of upsert, covering combinations of ({Height1, Height2}, Validator, {RepZero, Rep1, Rep2}).

@jayeshbairagi can you note this as a new ticket ? (either refer to this comment, or put it in more sensible wording :D )

File: test/validator/remove_validator.js [hunk 14]
```diff
@@ -27,6 +27,7 @@ contract('ValidatorSet::removeValidator', (accounts) => {
   let account;
   beforeEach(async () => {
     validatorSet = await ValidatorSet.new();
+    await validatorSet.setupValidator();
```
Comment (benjaminbollen) [inline test/validator/remove_validator.js]: why is the function not `setupValidatorSet()`

Comment (benjaminbollen) [reply test/validator/remove_validator.js]: (as a side note; this will conflict with the correction in #242)

**Review by benjaminbollen — DISMISSED** [review]
please move the `assert` up;
please have another look at my comment in `remove_validator.js`

File: test/self_protocore/upsert_validator.js (hunk 12 above)
Comment (benjaminbollen) [reply test/self_protocore/upsert_validator.js]: this can be improved later

Comment (benjaminbollen) [reply test/self_protocore/upsert_validator.js:40]: does not need to be done now; but add a @TODO, or @jayeshbairagi how should we track "known areas of improvements"

File: contracts/protocore/SelfProtocore.sol (hunk 9 above)
Comment (benjaminbollen) [reply contracts/protocore/SelfProtocore.sol]: yes, please make the ticket 💪

Im not resolving this because I think the `assert` should be at the beginning of the function

File: test/self_protocore/upsert_validator.js (hunk 12 above)
Comment (deepesh-kn) [reply test/self_protocore/upsert_validator.js]: Please take a reference from https://github.com/mosaicdao/mosaic-1/pull/244 so as to have minimum conflicts.

File: test/validator/remove_validator.js (hunk 14 above)
Comment (gulshanvasnani) [reply test/validator/remove_validator.js]: I have changed `setupValidator()` to `setupValidatorSetDouble()` after going through the PR #242

File: contracts/protocore/SelfProtocore.sol (hunk 11 above)
Comment (gulshanvasnani) [reply contracts/protocore/SelfProtocore.sol:99]: Done.

File: test/self_protocore/upsert_validator.js (hunk 12 above)
Comment (jayeshbairagi) [reply test/self_protocore/upsert_validator.js:40]: Done, noted for further discussion

File: test/self_protocore/upsert_validator.js (hunk 13 above)
Comment (jayeshbairagi) [reply test/self_protocore/upsert_validator.js:93]: Done, noted for further discussion.

File: contracts/protocore/SelfProtocore.sol (hunk 9 above)
Comment (gulshanvasnani) [reply contracts/protocore/SelfProtocore.sol]: Noted the point to make ticket.

**Review by deepesh-kn — APPROVED** [review]
Approving this PR with an assumption that the required checks will be added in the future as per @benjaminbollen implementation.
